File: 프로그래머스/lv2/김원래/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int visit[8];
int res = 0;
vector<vector<int>> dungeons;

void DFS(int curK, int dungeonCnt, int depth)
{
    if(depth == dungeons.size()) {
        res = max(res, dungeonCnt);
        return;
    }
    
    
    for(int cnt = 0 ; cnt < dungeons.size() ; ++cnt) {
        if(visit[cnt] != 1 && dungeons[cnt][0] <= curK) {
            visit[cnt] = 1;
            DFS(curK - dungeons[cnt][1], dungeonCnt + 1, depth +1);
            visit[cnt] = 0;
        }
        
        DFS(curK, dungeonCnt , depth + 1);
    }
}

int solution(int k, vector<vector<int>> _dungeons) {
    int answer = -1;
    
    dungeons = _dungeons;
    
    DFS(k, 0, 0);
    
    answer= res;
    
    return answer;
}


/*
k = 현재 유저 피로도
dungeons = 던전의 피로도 [최소 피로도, 소모 피로도]

*/

File: 프로그래머스/lv2/이재표/피로도.java
import java.util.*;
class Solution {
    int max=0;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        boolean[]visit=new boolean[dungeons.length];
        dfs(k,dungeons,visit,0);
        return max;
    }
    public void dfs(int k,int[][]dungeons,boolean[]visit,int cnt){
        for(int i=0;i<dungeons.length;i++){
            if(!visit[i] && k>=dungeons[i][0]){
                visit[i]=true;
                dfs(k-dungeons[i][1],dungeons,visit,cnt+1);
                visit[i]=false;
            }
        }
        max=Math.max(cnt,max);
        return;
    }
}
